atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2016 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.stream;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class BlockConsumingInputStream extends InputStream {
+    private final BlockGetter blockGetter;
+    private final long numBlocks;
+    private final int blocksInMemory;
+
+    private int nextBlockToRead;
+    private Iterator<Byte> buffer;
```

Comment:
this scares me from performance perspective in both iterating over single bytes as well as boxing overhead. Do we have any numbers on throughput for this?